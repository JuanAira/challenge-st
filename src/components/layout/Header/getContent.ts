interface Props {
  title: string,
  logo: { 
    text: string, 
  };
  links: {
    text: string,
  }[];
}

export const getContent = (): Props => {
  return {
    title: 'Test',
    logo: {
      text: 'Logo',
    },
    links: [
      {
        text: 'Link 1',
      },
      {
        text: 'Link 2',
      },
      {
        text: 'Link 3',
      },
      {
        text: 'Link 4',
      }
    ],
  }
}
