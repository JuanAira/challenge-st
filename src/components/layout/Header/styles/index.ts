import styled from 'styled-components';
import { UI } from '../../../../constants';

export const Container = styled.header<{ backgroundColor: string, mobile: boolean }>`
  height: ${UI.SIZE['GRID-72']};
  background-color: ${({ backgroundColor }) => backgroundColor};
  padding:  ${(props) => props.mobile ? UI.SIZE['GRID-4'] : UI.SIZE['GRID-20']};
  display: flex;
  justify-content: space-between;
  align-items: center;
  transition: all 0.4s ease-in;
`;

export const Title = styled.h1<{ color: string }>`
  color: ${({ color }) => color};
`;

export const SectionLinks = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

export const SectionTheme = styled.div`
  display: flex;
  align-items: center;
`;

export const DescriptionTheme = styled.p<{ color: string }>`
  color: ${({ color }) => color};
  margin: ${UI.SIZE['GRID-16']};
`;

export const DropdownContainer = styled.div<{ backgroundColor: string }>`
  display: flex;
  flex-direction: column;
  justify-content: center;
  background-color: ${({ backgroundColor }) => backgroundColor};
  align-items: center;
  transition: all 0.4s ease-in;

  & a:nth-child(1){
    margin-top: ${UI.SIZE['GRID-8']};
  }

  & a:nth-child(n){
    height: ${UI.SIZE['GRID-48']};
  }
`;

export const Link = styled.a<{ color: string }>`
  margin: 0 ${UI.SIZE['GRID-12']};
  color: ${({ color }) => color};
  cursor: pointer;
  transition: 0.4s;
  font-size: ${UI.FONT['SIZE-20']};

  &:hover:after {
    content: '';
    display: block;
    width: ${UI.SIZE['GRID-12']};
    border-bottom: 1px solid;
    margin: 0 auto;
  }
`;

export const Icon = styled.img`
  width: ${UI.SIZE['GRID-24']};
  height:${UI.SIZE['GRID-24']};
`;




