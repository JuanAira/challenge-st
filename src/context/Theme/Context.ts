/* eslint-disable @typescript-eslint/no-empty-function */
import React from 'react';
import { THEMES, THEME, UI } from '../../constants';

export interface ThemeDetails {
  themeName: THEMES;
  backgroundColor: string;
  color: string;
  description: string;
}

export const lightTheme: ThemeDetails = {
  themeName: THEMES.LIGHT,
  description: THEME.LIGHT.description,
  color: UI.COLORS.COLORS_THEME.DARK,
  backgroundColor: UI.COLORS.COLORS_THEME.LIGTH,
};

export const darkTheme: ThemeDetails = {
  themeName: THEMES.DARK,
  description: THEME.DARK.description,
  color: UI.COLORS.COLORS_THEME.LIGTH,
  backgroundColor: UI.COLORS.COLORS_THEME.DARK,
};

export default React.createContext({
  theme: darkTheme,
  isDarkTheme: false,
  toggleTheme: () => { },
});
