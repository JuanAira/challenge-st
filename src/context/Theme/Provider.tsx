import React, { useCallback, useEffect, useState } from 'react';
import { THEMES } from '../../constants';
import { ProviderProps } from './Provider.d';
import ThemeContext, {
  darkTheme,
  ThemeDetails,
  lightTheme,
} from './Context';

const Provider = ({ children }: ProviderProps) => {
  const isDarkThemeFromSystem = window.matchMedia(
    '(prefers-color-scheme: dark)',
  ).matches;

  const themeFromLocalStorage = window.localStorage.getItem('theme');

  const getThemeConfig = useCallback(() => {
    if (!themeFromLocalStorage) {
      if (isDarkThemeFromSystem) return darkTheme;
      return lightTheme;
    }
    if (themeFromLocalStorage === THEMES.DARK) return darkTheme;
    return lightTheme;
  }, [isDarkThemeFromSystem, themeFromLocalStorage]);

  const [theme, setTheme] = useState<ThemeDetails>(getThemeConfig());

  useEffect(() => {
    setTheme(getThemeConfig());
  }, [getThemeConfig]);

  const toggleTheme = () => {
    const themeToSet = isDarkTheme ? lightTheme : darkTheme;
    setTheme(themeToSet);
    window.localStorage.setItem('theme', themeToSet.themeName);
  };

  const isDarkTheme = theme.themeName === THEMES.DARK;
  const value = { theme, toggleTheme, isDarkTheme };

  return <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
};

export default Provider;
