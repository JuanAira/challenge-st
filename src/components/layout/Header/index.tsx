import React from 'react'
import { useContext, useState } from 'react';
import ThemeContext from '../../../context/Theme/Context';
import Switch from 'react-switch';
import useScreen from '../../../hooks/useScreen';
import { Button } from '../../../components/Button';
import { getContent } from './getContent';
import { THEME_IMAGE, UI } from '../../../constants';
import {
  Container,
  Title,
  SectionLinks,
  SectionTheme,
  Link,
  Icon,
  DescriptionTheme,
  DropdownContainer,
} from './styles';

export const Header = () => {
  const {
    theme: {
      themeName,
      backgroundColor,
      color,
      description,
    },
    isDarkTheme,
    toggleTheme,
  } = useContext(ThemeContext);
  const [dropdownLinks, setDropdownLinks] = useState<boolean>(false);
  const [isSwitchOn, setIsSwitchOn] = useState<boolean>(isDarkTheme ? true : false);
  const { links, title } = getContent();
  const isMobile = useScreen();

  const handleChangeThemeSwitch = (swicht: boolean) => {
    setIsSwitchOn(swicht);
    toggleTheme();
  };

  const renderAllLinks = () => {
    if (isMobile) { return <Button onClick={() => setDropdownLinks(!dropdownLinks)} />; }

    return (links.map(({ text }, key) => (
      <Link tabIndex={1} key={key} color={color}>
        {text}
      </Link>
    )));
  };

  return (
    <>
      <Container backgroundColor={backgroundColor} mobile={isMobile}>
        <Title color={color} data-testid="title">{title}</Title>

        <SectionLinks>{renderAllLinks()}</SectionLinks>

        <SectionTheme>
          <Icon src={THEME_IMAGE[themeName].img} alt={THEME_IMAGE[themeName].alt} />

          {!isMobile && (<DescriptionTheme color={color}>{description}</DescriptionTheme>)}

          <Switch
            checked={isSwitchOn}
            onChange={handleChangeThemeSwitch}
            offHandleColor={UI.COLORS['PURPLE-500']}
            onColor={UI.COLORS['PURPLE-500']}
            onHandleColor={UI.COLORS['PURPLE-100']}
            handleDiameter={20}
            uncheckedIcon={false}
            checkedIcon={false}
            height={14}
            width={36}
            id="switch"
          />
        </SectionTheme>
      </Container>

      {dropdownLinks && isMobile && (
        <DropdownContainer backgroundColor={backgroundColor}>
          {links.map((link, key) => (<Link key={key} color={color}>{link.text}</Link>))}
        </DropdownContainer>
      )}
    </>
  );
};

export default Header;
