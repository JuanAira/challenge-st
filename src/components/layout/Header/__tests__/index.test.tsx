import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import ThemeProvider from '../../../../context/Theme/Provider';
import ThemeContext from '../../../../context/Theme/Context';
import { Header } from '../../Header';
import { lightTheme } from '../../../../context/Theme/Context';

describe('Header component', () => {
  beforeAll(() => {
    window.matchMedia = window.matchMedia || function () {
      return {
        matches: false,
        addListener: jest.fn(),
        removeListener: jest.fn(),
      };
    };
  });

  test('should match to snapshot - base - dark', () => {
    const { container } = render(
      <ThemeProvider>
        <Header />
      </ThemeProvider>
    );
    expect(container).toMatchSnapshot();
  });

  test('should match to snapshot - ligth', () => {
    const value = {
      theme: lightTheme,
      toggleTheme: jest.fn(),
      isDarkTheme: true,
    };

    const { container } = render(
      <ThemeContext.Provider value={value}>
        <Header />
      </ThemeContext.Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test('should match to snapshot - ligth - mobile', () => {
    global.innerWidth = 200;
    global.dispatchEvent(new Event('resize'));

    const value = {
      theme: lightTheme,
      toggleTheme: jest.fn(),
      isDarkTheme: true,
    };

    const { container } = render(
      <ThemeContext.Provider value={value}>
        <Header />
      </ThemeContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });

  test('should to macht the correct links - mobile', () => {
    global.innerWidth = 200;
    global.dispatchEvent(new Event('resize'));

    const { container } = render(
      <ThemeProvider>
        <Header />
      </ThemeProvider>
    );
    const links = container.querySelectorAll('a');
    expect(links.length).toBe(0);
  });

  test('should to macht the correct links - desktop', () => {
    global.innerWidth = 1200;
    global.dispatchEvent(new Event('resize'));

    const { container } = render(
      <ThemeProvider>
        <Header />
      </ThemeProvider>
    );
    const links = container.querySelectorAll('a');
    expect(links.length).toBe(4);
  });

  test('should to call the fn', () => {
    const value = {
      theme: lightTheme,
      toggleTheme: jest.fn(),
      isDarkTheme: true,
    };

    const { container } = render(
      <ThemeContext.Provider value={value}>
        <Header />
      </ThemeContext.Provider>
    );
    const swicht = container.querySelectorAll('input');
    fireEvent.click(swicht[0]);
    expect(value.toggleTheme).toHaveBeenCalledTimes(1);
  });

  test('should match to snapshot - mobile - dark', async () => {
    global.innerWidth = 200;
    global.dispatchEvent(new Event('resize'));

    const { container } = render(
      <ThemeProvider>
        <Header />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
