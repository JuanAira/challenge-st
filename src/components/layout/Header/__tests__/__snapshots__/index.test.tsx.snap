// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should match to snapshot - base - dark 1`] = `
<div>
  <header
    class="sc-dkzDqf iLFJD"
  >
    <h1
      class="sc-hKMtZM gTMcNY"
      color="#171B23"
      data-testid="title"
    >
      Test
    </h1>
    <div
      class="sc-eCYdqJ RkdUt"
    >
      <a
        class="sc-ftvSup ebRLgE"
        color="#171B23"
        tabindex="1"
      >
        Link 1
      </a>
      <a
        class="sc-ftvSup ebRLgE"
        color="#171B23"
        tabindex="1"
      >
        Link 2
      </a>
      <a
        class="sc-ftvSup ebRLgE"
        color="#171B23"
        tabindex="1"
      >
        Link 3
      </a>
      <a
        class="sc-ftvSup ebRLgE"
        color="#171B23"
        tabindex="1"
      >
        Link 4
      </a>
    </div>
    <div
      class="sc-jSMfEi dnCBFt"
    >
      <img
        alt="light theme"
        class="sc-papXJ gwGOaF"
        src="night_night_mode_moon_crescent_dark_mode_icon.svg"
      />
      <p
        class="sc-gKXOVf eZouWf"
        color="#171B23"
      >
        Light mode
      </p>
      <div
        style="position: relative; display: inline-block; text-align: left; opacity: 1; direction: ltr; border-radius: 7px; transition: opacity 0.25s; user-select: none;"
      >
        <div
          class="react-switch-bg"
          style="height: 14px; width: 36px; margin: 3px; position: relative; background: rgb(136, 136, 136); border-radius: 7px; cursor: pointer; transition: background 0.25s;"
        />
        <div
          class="react-switch-handle"
          style="height: 20px; width: 20px; background: rgb(91, 118, 177); display: inline-block; cursor: pointer; border-radius: 50%; position: absolute; transform: translateX(0px); top: 0px; outline: 0; border: 0px; transition: background-color 0.25s, transform 0.25s, box-shadow 0.15s;"
        />
        <input
          aria-checked="false"
          id="switch"
          role="switch"
          style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
          type="checkbox"
        />
      </div>
    </div>
  </header>
</div>
`;

exports[`Header component should match to snapshot - ligth - mobile 1`] = `
<div>
  <header
    class="sc-dkzDqf flCaaF"
  >
    <h1
      class="sc-hKMtZM gTMcNY"
      color="#171B23"
      data-testid="title"
    >
      Test
    </h1>
    <div
      class="sc-eCYdqJ RkdUt"
    >
      <button
        class="sc-bczRLJ dlyGG"
      >
        <img
          alt="light theme hamburger"
          class="sc-gsnTZi ccUHbM"
          src="elements_hamburger_interface_list_menu_ui_icon_light.svg"
        />
      </button>
    </div>
    <div
      class="sc-jSMfEi dnCBFt"
    >
      <img
        alt="light theme"
        class="sc-papXJ gwGOaF"
        src="night_night_mode_moon_crescent_dark_mode_icon.svg"
      />
      <div
        style="position: relative; display: inline-block; text-align: left; opacity: 1; direction: ltr; border-radius: 7px; transition: opacity 0.25s; user-select: none;"
      >
        <div
          class="react-switch-bg"
          style="height: 14px; width: 36px; margin: 3px; position: relative; background: rgb(91, 118, 177); border-radius: 7px; cursor: pointer; transition: background 0.25s;"
        />
        <div
          class="react-switch-handle"
          style="height: 20px; width: 20px; background: rgb(117, 152, 228); display: inline-block; cursor: pointer; border-radius: 50%; position: absolute; transform: translateX(22px); top: 0px; outline: 0; border: 0px; transition: background-color 0.25s, transform 0.25s, box-shadow 0.15s;"
        />
        <input
          aria-checked="true"
          checked=""
          id="switch"
          role="switch"
          style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
          type="checkbox"
        />
      </div>
    </div>
  </header>
</div>
`;

exports[`Header component should match to snapshot - ligth 1`] = `
<div>
  <header
    class="sc-dkzDqf iLFJD"
  >
    <h1
      class="sc-hKMtZM gTMcNY"
      color="#171B23"
      data-testid="title"
    >
      Test
    </h1>
    <div
      class="sc-eCYdqJ RkdUt"
    >
      <a
        class="sc-ftvSup ebRLgE"
        color="#171B23"
        tabindex="1"
      >
        Link 1
      </a>
      <a
        class="sc-ftvSup ebRLgE"
        color="#171B23"
        tabindex="1"
      >
        Link 2
      </a>
      <a
        class="sc-ftvSup ebRLgE"
        color="#171B23"
        tabindex="1"
      >
        Link 3
      </a>
      <a
        class="sc-ftvSup ebRLgE"
        color="#171B23"
        tabindex="1"
      >
        Link 4
      </a>
    </div>
    <div
      class="sc-jSMfEi dnCBFt"
    >
      <img
        alt="light theme"
        class="sc-papXJ gwGOaF"
        src="night_night_mode_moon_crescent_dark_mode_icon.svg"
      />
      <p
        class="sc-gKXOVf eZouWf"
        color="#171B23"
      >
        Light mode
      </p>
      <div
        style="position: relative; display: inline-block; text-align: left; opacity: 1; direction: ltr; border-radius: 7px; transition: opacity 0.25s; user-select: none;"
      >
        <div
          class="react-switch-bg"
          style="height: 14px; width: 36px; margin: 3px; position: relative; background: rgb(91, 118, 177); border-radius: 7px; cursor: pointer; transition: background 0.25s;"
        />
        <div
          class="react-switch-handle"
          style="height: 20px; width: 20px; background: rgb(117, 152, 228); display: inline-block; cursor: pointer; border-radius: 50%; position: absolute; transform: translateX(22px); top: 0px; outline: 0; border: 0px; transition: background-color 0.25s, transform 0.25s, box-shadow 0.15s;"
        />
        <input
          aria-checked="true"
          checked=""
          id="switch"
          role="switch"
          style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
          type="checkbox"
        />
      </div>
    </div>
  </header>
</div>
`;

exports[`Header component should match to snapshot - mobile - dark 1`] = `
<div>
  <header
    class="sc-dkzDqf flCaaF"
  >
    <h1
      class="sc-hKMtZM gTMcNY"
      color="#171B23"
      data-testid="title"
    >
      Test
    </h1>
    <div
      class="sc-eCYdqJ RkdUt"
    >
      <button
        class="sc-bczRLJ dlyGG"
      >
        <img
          alt="light theme hamburger"
          class="sc-gsnTZi ccUHbM"
          src="elements_hamburger_interface_list_menu_ui_icon_light.svg"
        />
      </button>
    </div>
    <div
      class="sc-jSMfEi dnCBFt"
    >
      <img
        alt="light theme"
        class="sc-papXJ gwGOaF"
        src="night_night_mode_moon_crescent_dark_mode_icon.svg"
      />
      <div
        style="position: relative; display: inline-block; text-align: left; opacity: 1; direction: ltr; border-radius: 7px; transition: opacity 0.25s; user-select: none;"
      >
        <div
          class="react-switch-bg"
          style="height: 14px; width: 36px; margin: 3px; position: relative; background: rgb(136, 136, 136); border-radius: 7px; cursor: pointer; transition: background 0.25s;"
        />
        <div
          class="react-switch-handle"
          style="height: 20px; width: 20px; background: rgb(91, 118, 177); display: inline-block; cursor: pointer; border-radius: 50%; position: absolute; transform: translateX(0px); top: 0px; outline: 0; border: 0px; transition: background-color 0.25s, transform 0.25s, box-shadow 0.15s;"
        />
        <input
          aria-checked="false"
          id="switch"
          role="switch"
          style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
          type="checkbox"
        />
      </div>
    </div>
  </header>
</div>
`;
